Add documentation for helm memberlist cluster label migration (#7961)
grafana/mimir: docs/sources/helm-charts/mimir-distributed/migration-guides/migrate-enable-cluster-label-verification/_index.md

  different Memberlist cluster.
menuTitle: "Configure a unique Memberlist cluster label"
title: "Configure a unique Grafana Mimir's Memberlist cluster label in the mimir-distributed Helm chart installation"
weight: 110
---

# Configure a unique Grafana Mimir's Memberlist cluster label in the mimir-distributed Helm chart installation

This document shows the step to configure cluster label verification in a Grafana Mimir installed by Helm.
Multiple [Memberlist](https://grafana.com/docs/mimir/<MIMIR_VERSION>/references/architecture/memberlist-and-the-gossip-protocol/) [gossip ring](https://grafana.com/docs/mimir/<MIMIR_VERSION>/references/architecture/hash-ring/) cluster is at risk of merging into one without enabling cluster label verification.
For example, if a Mimir, Tempo or Loki are running in the same Kubernetes cluster, they might communicate with each other without this configuration update.
Once cluster label verification is enabled, before Mimir components communicate with other components, it will verify whether those other components are having same cluster label.

> The process to update the configuration should take around 30 minutes.

Author: I think this make more sense. Even 30 minutes that I put there was also quite abrupt.

> The process to update the configuration should take around 30 minutes.

Reviewer: did you mean to apply this change to the doc?

> The process to update the configuration should take around 30 minutes.

Author: Strange. I remember that I changed this as what was suggested. Maybe it got stashed in between of other commits. I confirm the commit this time in cf11b13756.

## Before you begin

- You have a Grafana Mimir installed by mimir-distributed helm chart with its Memberlist cluster label still set to default value.
- You have `kubectl` and `helm` command line configured to connect to the Kubernetes cluster where your Grafana Mimir is running.

## Configuration update steps

There are three steps of the configuration update:

1. Disable Memberlist cluster label verification
1. Set cluster label to all Mimir components
1. Enable Memberlist cluster label verification again

### 1. Disable Memberlist cluster label verification

Cluster label verification flag is enabled by default with cluster label set to an empty string.

> Letting default value of cluster label can make different systems that use Memberlist to communicate with each other if they also have not updated the default cluster label.

Reviewer: ```suggestion
Using the default value of cluster label can make different systems that use Memberlist communicate with each other if they also have not updated the default cluster label.
```

Updating a new cluster label directly to a non-empty string value without disabling cluster label verification will cause Memberlist to form partition in the Grafana Mimir cluster.
The partition makes some Mimir components to have different cluster label values which can make the cluster member can't communicate.
To disable cluster label verification flag, set the following structured config in mimir-distributed values.yaml configuration.

```yaml
mimir:
  structuredConfig:
    memberlist:
      cluster_label_verification_disabled: true
```

> Rollout the installation to apply the configuration changes by running `helm upgrade <my-mimir-release> mimir-distributed -f values.yaml`. Replace `my-mimir-release` with the actual Mimir release name.

Reviewer: Also perhaps worth noting that they should wait until all Pods are ready. Here and after step 2

> ### 2. Set cluster label to all Mimir components

Reviewer: ```suggestion
### 2. Set cluster label on all Mimir components
```

Set cluster label to all Mimir components by setting the following configuration below.
The configuration will set `cluster_label` to the Helm release name and the namespace where the helm release is installed.

> Updating a new cluster label after disabling cluster label verification will prevent Memberlist to form partition.

Reviewer: ```suggestion
Updating a new cluster label after disabling cluster label verification will prevent Memberlist from forming a partition.
```

```yaml
mimir:
  structuredConfig:
    memberlist:
      cluster_label_verification_disabled: true
      cluster_label: "{{.Release.Name}}-{{.Release.Namespace}}"
```

Apply the configuration changes again by running `helm upgrade <my-mimir-release> mimir-distributed -f values.yaml`.

### 3. Enable Memberlist cluster label verification

Remove `mimir.structuredConfig.memberlist.cluster_label_verification_disabled` from the values.yaml file to re-enable Memberlist cluster label verification.

```yaml
mimir:
  structuredConfig:
    memberlist:
      cluster_label: "{{.Release.Name}}-{{.Release.Namespace}}"
```

Apply the configuration changes by running `helm upgrade <my-mimir-release> mimir-distributed -f values.yaml`.

## Verifying The Migration

Once the rollout is completed, verify the change by looking at the `/memberlist` endpoint in some of Grafana Mimir pods.
Run the following port-forward command on several different Grafana Mimir components.

```bash

>    kubectl port-forward pod/<mimir-pod-1> --kube-context=<my-k8s-context> --namespace=<my-mimir-namespace> 8080:80
>    kubectl port-forward pod/<mimir-pod-2> --kube-context=<my-k8s-context> --namespace=<my-mimir-namespace> 8081:80

Reviewer: the helm chart deploys pods listening on port 8080 by default
```suggestion
   kubectl port-forward pod/<mimir-pod-1> --kube-context=<my-k8s-context> --namespace=<my-mimir-namespace> 8080:8080
   kubectl port-forward pod/<mimir-pod-2> --kube-context=<my-k8s-context> --namespace=<my-mimir-namespace> 8081:8080
```

```

Replace mimir-pod with several actual pods from different Mimir components.
Ensure the host port 8080 and 8081 are available, otherwise use different available ports.

> Make sure that the container port which is set by default to port 80 is also correct.

Reviewer: can perhaps drop this since Mimir is listening on 8080 in the helm chart

---

> description: "Migrating Mimir Helm Installation to Use Memberlist Cluster Label"

Reviewer 2: I think the description should provide more information on what the reader can expect to get from this page.
The description is presented to the user in search engines and on social media when links to this page are used.

A good example is https://github.com/grafana/mimir/blob/main/docs/sources/mimir/set-up/migrate/migrate-from-cortex.md?plain=1#L4-L7

https://grafana.com/docs/writers-toolkit/write/front-matter/#description-required

> title: "Migrating Mimir Helm Installation to Use Memberlist Cluster Label"
> menuTitle: "Migrating Mimir Helm Installation to Use Memberlist Cluster Label"

Reviewer 2: Titles and headings should use sentence case rather than title case.
We should also avoid gerunds ("ing" endings in verbs).
Titles should also use the full product name "Grafana Mimir" for search engine optimization.

I have corrected this as an example, but please see my general review comment because we may want to rename this page.

```suggestion
title: "Migrate Grafana Mimir Helm chart to use memberlist cluster label"
```

Also the `menuTitle` can be omitted if it is the same as the title.
In this case, I think we need a much shorter title that omits the product name.
Considering a page rename may make this `menuTitle` much shorter, perhaps `Configure a memberlist cluster label`?

weight: 110
---

# Migrating Mimir Helm Installation to Use Memberlist Cluster Label

> ## Introduction

Reviewer 2: In general, we don't want stacked headings (headings with no content between them).

The introductory content can be moved out of the H2 and become the conceptual material at the start of this page.

> Grafana Mimir uses [memberlist] to share works and deciding which component replica to send the workload such as when ingesting series. Memberlist encodes the replica information in data structure called [hash ring], this is a consistent hashing scheme in which different ingesters instance tokens are placed around the ring. The token position in the ring determines which ingester should handle a request. The information of this hash ring is delivered to different components by using gossip protocol.

Author: This section has been shortened and I tried to update the paragraph to make it more accurate.

> By default, hash ring is global. If we also have Grafana Tempo and Grafana Loki in the same Kubernetes cluster, their components can unintentionally talk and sending data with each other, because Loki and Tempo also uses memberlist. This possibility to happen increases in cluster where the pods may reuse each other's IPs after churning.

Reviewer: that first sentence is a bit unclear to me. I think saying that the key-value store where the hash ring is stored is viewed by all components.

re second sentence: I'd mention that loki and tempo clusters can end up participating in the same memberlist cluster and their key-value stores can merge, which in turn can lead to components from the different clusters to try to communicate with each other (I'd prefer "communicate" over "talk")

also re styling: I think "pod" should be "Pod" because that's the naming that k8s uses in its docs

Author: Most of this paragraph is removed. Let me know if you think we should keep information about the possible Pod reuse after churning.

In this document we will describe the step on how to migrate a Mimir installation to prevent two separate memberlist gossip hash ring to join into one.

There are three steps of the migration which we will describe in details in the migration section. But in brief summary the steps are:

1. Disable memberlist cluster label verification
1. Set cluster label to all Mimir components
1. Enable memberlist cluster label verification again

> The migration should take around 30 minutes. The risk of not doing this migration if you have Mimir and Tempo or Loki in the same cluster is the possibility of those different backend to talk with each other.

Reviewer 2: I think the second sentence on L24 is the most important bit of this introduction. A reader wants to know why they should perform this task and they may not really care about the conceptual information on L12.

Author: I removed many repetition and make a more concise introduction in the beginning of the section.

## How The Migration Solve The Issue

> Non-Mimir component can merge with Mimir memberlist because by default, memberlist is not namespaced and applies globally. As an example, consider a Loki pod is terminated and the IP address reused by a new Mimir pod in the same cluster. This might cause other Loki components try to talk and sending data to this new Mimir pod.

Reviewer: This looks the same as  the paragraph on line 14. I'd perhaps remove that one and only keep the paragraph in this section

Author: This section is removed.

Cluster label will prevent such situation by only allowing communication for components that has same cluster label. Once that is enabled, before memberlist try to communicate with other components, it will verify whether that new components are having same cluster label and only allowing memberlist to communicate if the parties are having the same cluster label.

## Migration

### 1. Disable memberlist cluster label verification

> Cluster label verification flag is enabled by default. However, cluster label default value is an empty string. If different systems that using memberlist and not setting the cluster label, they can talk with each other. To disable cluster label verification flag, set the following structured config.

Reviewer: do we want to mention why we are doing this step? It's because if we just change the cluster label without first disabling it, then we'd get a memberlist partition in the mimir cluster: pods will only be able to communicate over memberlist with pods with the same cluster label. I think some brief context might help. Not super sure though

Author: I added the context. Let me know if that can be better.

> ```
> mimir

Reviewer: ```suggestion
```yaml
mimir:
```

Reviewer: can you apply the same in the examples below?

  structuredConfig:
    memberlist:
      cluster_label_verification_disabled: true
```

> Rollout the installation to apply the configuration changes by running `helm upgrade mimir-distributed -f values.yaml`.

Reviewer: does this command work for you? I think helm upgrade needs two args this works for me

```bash
helm upgrade mimir grafana/mimir-distributed -f values.yaml
```

Reviewer: i think the commands below also need changing

### 2. Set cluster label to all Mimir components

Set cluster label to all Mimir components by setting the following configuration. Once the configuration is applied, all Mimir components will only communicate via memberlist if the other component is having the same cluster label.

```
mimir
  structuredConfig:
    memberlist:
      cluster_label_verification_disabled: true

>       cluster_label: "$helm-release-name"

Reviewer: we can use a go template here so different mimir clusters in the same cluster don't mix up. And also don't require users to change the yaml

```
cluster_label: '{{.Release.Name}}-{{.Release.Namespace}}'
```

may be worth explaining to the audience what this template would do though

Author: Updated the template and some brief explanation about it.

```

Apply the configuration changes by running `helm upgrade mimir-distributed -f values.yaml`.

### 3. Enable memberlist cluster label verification

> Set `memberlist.cluster_label_verification_disabled` to false, to re-enable memberlist cluster label verification.

Reviewer: technically the path is `mimir.structuredConfig.memberlist.cluster_label_verification_disabled`

Reviewer: also can't we just remove the key from the config? That way users don't need to have some line in their config "forever"

Author: Removed the config and mentioned it in the text that the removal of the key config means re-enabling cluster label verification with the new cluster label.

```
mimir
  structuredConfig:
    memberlist:
      cluster_label_verification_disabled: false
      cluster_label: "$helm-release-name"
```

Apply the configuration changes by running `helm upgrade mimir-distributed -f values.yaml`.

## Verifying The Migration

Once the rollout has been done run the following helm command to verify if we have successfully migrated the Mimir memberlist to check the cluster label.

```
helm --kube-context=[your-k8s-context] --namespace=[your-mimir-namespace] get values [your-mimir-release-name]
```

You should see the following config the in the values as what we have set above.

```
mimir
  structuredConfig:
    memberlist:
      cluster_label_verification_disabled: false
      cluster_label: "$helm-release-name"

> ```

Reviewer: Let's get folks to visit the `/memberlist` endpoint on a couple of pods to make sure they have the same view of all members. I think that would be more convincing verification because it means components are actually part of the same memberlist cluster.

Author: Didn't think about `/memberlist` endpoint. Yes that is make more sense to use this for the verification instead of just printing out back the values.yaml.

{{% docs/reference %}}
[memberlist]: "/ -> /docs/mimir/<MIMIR_DOCS_VERSION>/references/architecture/memberlist-and-the-gossip-protocol"
[hash ring]: "/ -> /docs/mimir/<MIMIR_DOCS_VERSION>/references/architecture/hash-ring"

> {{% /docs/reference %}}

Reviewer 2: I don't think there is any need to use `docs/reference` here because the links are the same and this content isn't reused via Hugo mounts as far as I know. You only need to use that shortcode for [content that is used in multiple website projects](https://grafana.com/docs/writers-toolkit/write/links/#link-from-source-content-thats-used-in-multiple-projects) like being both in Grafana and Grafana Cloud documentation.

Instead, you can just use URLs with `<MIMIR_DOCS_VERSION>` for version substitution.

```

Apply the configuration changes again by running `helm upgrade <my-mimir-release> mimir-distributed -f values.yaml`.

> Wait until all Pods are ready before going to the next step.

Reviewer: ```suggestion
Replace `<my-mimir-release>` with the actual Mimir release name. Wait until all Pods are ready before going to the next step.
```

```

Apply the configuration changes by running `helm upgrade <my-mimir-release> mimir-distributed -f values.yaml`.

> Wait until all Pods are ready before verifying that the configuration is applied correctly.

Reviewer: ```suggestion
Replace `<my-mimir-release>` with the actual Mimir release name. Wait until all Pods are ready before verifying that the configuration is applied correctly.
```

## Verifying the configuration changes

Once the rollout is completed, verify the change by looking at the `/memberlist` endpoint in some of Grafana Mimir pods.
Run the following port-forward command on several different Grafana Mimir components.

```bash
   kubectl port-forward pod/<mimir-pod-1> --kube-context=<my-k8s-context> --namespace=<my-mimir-namespace> 8080:8080
   kubectl port-forward pod/<mimir-pod-2> --kube-context=<my-k8s-context> --namespace=<my-mimir-namespace> 8081:8080
```

> Replace <mimir-pod-1> and <mimir-pod-2> with several actual pods from different Mimir components.

Reviewer: ```suggestion
Replace `<mimir-pod-1>` and `<mimir-pod-2>` with several actual pods from different Mimir components.
```

I think that's done in the other places that use the angled brackets
